drv/lpc55-swd/src/main.rs

```diff
@@ -0,0 +1,656 @@
+const DP_CTRL_CDBGPWRUPACK: u32 = 1 << 29;
+
+// See Ch5 of ARM ADI for bit pattern
+const START_BIT: u8 = 7;
+// Stop is bit 1 and always 0
+const PARITY_BIT: u8 = 2;
+const ADDR_BITS: u8 = 3;
+
+const RDWR_BIT: u8 = 5;
+const APDP_BIT: u8 = 6;
+const PARK_BIT: u8 = 0;
+
+const START_VAL: u8 = 1 << START_BIT;
+const PARK_VAL: u8 = 1 << PARK_BIT;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Port {
+    DP = 0,
+    AP = 1,
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialEq)]
+enum DpRead {
+    IDCode = 0x0,
+    Ctrl = 0x4,
+    //Resend = 0x8,
+    //Rdbuf = 0xc,
+}
+
+impl DpRead {
+    fn addr_bits(&self) -> u8 {
+        // we need this to be as bits [2:3]
+        match *self {
+            DpRead::IDCode => 0b00,
+            DpRead::Ctrl => 0b10,
+        }
+    }
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialEq)]
+enum DpWrite {
+    Abort = 0x0,
+    Ctrl = 0x4,
+    Select = 0x8,
+}
+
+impl DpWrite {
+    fn addr_bits(&self) -> u8 {
+        // we need this to be as bits [2:3]
+        match *self {
+            DpWrite::Abort => 0b00,
+            DpWrite::Ctrl => 0b10,
+            DpWrite::Select => 0b01,
+        }
+    }
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialEq)]
+enum RawSwdReg {
+    DpRead(DpRead),
+    DpWrite(DpWrite),
+    ApRead(ApReg),
+    ApWrite(ApReg),
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialEq)]
+enum ApReg {
+    CSW = 0x0,
+    TAR = 0x4,
+    DRW = 0xC,
+    //BD0 = 0x10,
+    //BD1 = 0x14,
+    //BD2 = 0x18,
+    //BD3 = 0x1C,
+    //ROM = 0xF8,
+    IDR = 0xFC,
+}
+
+impl ApReg {
+    fn addr_bits(&self) -> u8 {
+        // we need this to be as bits [2:3]
+        match *self {
+            ApReg::CSW => 0b00,
+            ApReg::TAR => 0b10,
+            ApReg::DRW => 0b11,
+
+            ApReg::IDR => 0b11,
+        }
+    }
+}
+
+// represents the port + register
+struct ApAddr(u32, ApReg);
+
+fn get_addr_and_rw(reg: RawSwdReg) -> (u8, u8) {
+    match reg {
+        RawSwdReg::DpRead(v) => (1 << RDWR_BIT, v.addr_bits() << ADDR_BITS),
+        RawSwdReg::DpWrite(v) => (0 << RDWR_BIT, v.addr_bits() << ADDR_BITS),
+        RawSwdReg::ApRead(v) => (1 << RDWR_BIT, v.addr_bits() << ADDR_BITS),
+        RawSwdReg::ApWrite(v) => (0 << RDWR_BIT, v.addr_bits() << ADDR_BITS),
+    }
+}
+
+// The parity is only over 4 of the bits
+fn calc_parity(val: u8) -> u8 {
+    let b = val >> 3 & 0xf;
+
+    if b.count_ones() % 2 == 0 {
+        0 << PARITY_BIT
+    } else {
+        1 << PARITY_BIT
+    }
+}
+
+struct ServerImpl {
+    spi: spi_core::Spi,
+    init: bool,
+}
+
+impl idl::InOrderSpCtrlImpl for ServerImpl {
+    fn read(
+        &mut self,
+        _: &RecvMessage,
+        addr: u32,
+        dest: LenLimit<Leased<W, [u8]>, 4096>,
+    ) -> Result<(), RequestError<SpCtrlError>> {
+        ringbuf_entry!(Trace::ReadCmd);
+        if !self.init {
+            return Err(SpCtrlError::NeedInit.into());
+        }
+        let cnt = dest.len();
+        if cnt % 4 != 0 {
```

Comment:
One option to satisfy these properties statically, and make your code below simpler, is to use a lease of `u32` instead of `u8`.

Reply:
It turns out this doesn't work out of the box:
```
error[E0308]: mismatched types
  --> /hubris/target/thumbv8m.main-none-eabihf/release/build/drv-sp-ctrl-api-7da50a9d9a750d44/out/client_stub.rs:61:44
   |
61 |                 userlib::Lease::write_only(arg_sink),
   |                                            ^^^^^^^^ expected `u8`, found `u32`
   |
   = note: expected mutable reference `&mut [u8]`
              found mutable reference `&mut [u32]`
```
Leases work on `[u8]` at the moment. I'm going to leave this as is for now.